flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -3478,6 +3481,45 @@ class ArrayExprLowering {
     auto idxTy = builder.getIndexType();
     auto one = builder.createIntegerConstant(loc, idxTy, 1);
 
+    if (!eleSz) {
+      // Compute the element size at runtime.
+      assert(fir::hasDynamicSize(eleTy));
+      if (auto charTy = eleTy.dyn_cast<fir::CharacterType>()) {
+        auto charBytes =
+            builder.getKindMap().getCharacterBitsize(charTy.getFKind()) / 8;
+        auto bytes = builder.createIntegerConstant(loc, idxTy, charBytes);
+        auto length = fir::getLen(exv);
```

Comment:
I think characters expressions in the array constructor do not have to have the same length as the array-ctor elements after F95. So if the length is not the same, there may be some padding/truncating needed, and the expression lengths may not be usable for the memcpy. I did not test, but I suspect the following program might not work as intended (unless, SetLength are inserted by the frontend, but then that would most likely create a intermediate temp).

```
subroutine test(c1, c2, c3)
  character(10) :: c1
  character(5) :: c2
  character(8) :: c3(2)
  ! Need to truncate c1 and pad c2 when copying them in the array-ctor
  c3 = (/character(8)::c1, c2/)
end
```

So I think you should maybe consider using the same logic as scalar and character assignments here.